rfc/system/RFC-0000-execution-concurrency.md

```diff
@@ -0,0 +1,232 @@
+
+In models/execution.go:
+```go
+
+type Execution struct {
+	
+	...
+	// Already exists
+	LaunchPlanID uint   `gorm:"index"`
+	// New field to make querying on the named entity 
+	LaunchPlanNamedEntityID uint `gorm:"index"`
+	
+}
+
+```
+
+Then the reconciliation loop would query executions in a non-terminal phase matching the launch plan named entity ID instead of LaunchPlanID.
+
+## 4 Metrics & Dashboards
+- Time spent in PENDING: It's useful to understand the duration spent in PENDING before a launch plan transitions to RUNNING
+- It may be useful for Flyte platform operators to also configure alerts if an execution stays in PENDING for too long of a threshold
+
+## 5 Drawbacks
+
+The [executions model](https://github.com/flyteorg/flyte/blob/38883c721dac2875bdd2333f4cd56e757e81ea5f/flyteadmin/pkg/repositories/models/execution.go)
+already has indices on
+- primary key index
+- launch plan id
+- workflow id
+- task id (for single task executions)
+- execution created at
+- error kind (code)
+- user who launched the execution
+- state
+
+Database performance suffers as new indices are added (ref [[1](https://use-the-index-luke.com/sql/dml/insert)] [[2](https://www.timescale.com/learn/postgresql-performance-tuning-optimizing-database-indexes)])
+We could as an alternative, repurpose the existing launch plan index to include (launch plan id, phase, created at) to optimize the query for pending executions and not significantly affect queries on launch plan id leveraging the existing index.
+
+
+## 6 Alternatives
+
+### Scheduling
+This proposal purposefully uses FIFO scheduling but there is a chance we may want to define other scheduling orders or catch-up policies.
+
+To accomplish this, we can extend the `ConcurrenyPolicy` proto message to encapsulate scheduling behavior
+
+```protobuf
+message Concurrency {
+  // Defines how many executions with this launch plan can run in parallel
+  uint32 max = 1;
+  
+  // Defines how to handle the execution when the max concurrency is reached.
+  ConcurrencyPolicy policy = 2;
+
+  ConcurrencyScheduling scheduling = 3;
+}
+
+
+type ConcurrencyScheduling enum {
+  FIFO = 0;
+  FILO = 1;
+  ...
+}
+```
+
+Furthermore, we may want to introduce a max pending period to fail executions that have been in `PENDING` for too long
+
+
+## 7 Potential Impact and Dependencies
+
+*Here, we aim to be mindful of our environment and generate empathy towards others who may be impacted by our decisions.*
+
+- *What other systems or teams are affected by this proposal?*
+- *How could this be exploited by malicious attackers?*
+
+## 8 Unresolved questions
+
+- Should we always attempt to schedule pending executions in ascending order of creation time?
+    - Decision: We'll use FIFO scheduling by default but can extend scheduling behavior with an enum going forward.
```

Comment:
If this has been decided (I'm ok with it), could you please reformulate in the text above where this is still discussed as an open question? 🙏

Reply:
updated the discussion above!